File: chapter01/02/prog.rb
# ruby prog.rb

def compose(g, f)
  -> (a) { g.call(f.call(a)) }
end

is_empty = -> (len) { len == 0 }
length = -> (str) { str.length }

is_empty_string = compose(is_empty, length)

puts is_empty_string.call("")
puts is_empty_string.call("nest")

File: chapter01/02/prog.cpp
// $ g++ prog.cpp
// $ ./a.out
#include <iostream>
#include <string>
using namespace std;

template <class A, class B, class C>
auto compose(C g(B), B f(A)) {
  return [=](A a) {
    return g(f(a));
  };
}

int length(string s) { 
  return s.length();
}

bool isEmpty(int length) {
  return length == 0;
}

int main() {
  auto isEmptyString = compose(isEmpty, length);
  cout << isEmptyString("") << endl;
  cout << isEmptyString("nest") << endl;

  return 0;
}